File: Interviews/LeetCode/Daily-Problems/2022/October/Day-16/1335.py
class Solution:
    def dp(self, idx, d, maxWork, jd, cache) -> int:
        if idx == len(jd):
            if d > 0: return 1e9
            return maxWork
        if cache[idx][d][maxWork] != -1:
            return cache[idx][d][maxWork]
        
        ans = self.dp(idx + 1, d, max(maxWork, jd[idx]), jd, cache)
        if d > 0:
            ans = min(ans, self.dp(idx + 1, d - 1, jd[idx], jd, cache) + maxWork)
            
        cache[idx][d][maxWork] = ans
        return ans
    
    def minDifficulty(self, jD: List[int], d: int) -> int:
        n = len(jD)
        cache = [[[-1 for i in range(1001)] for j in range(d)] for k in range(n)]
        ans = self.dp(1, d - 1, jD[0], jD, cache)
        if ans == 1e9:
            return -1
        return ans

File: Interviews/LeetCode/Topic-Wise/Dynamic-Programming/Hard/Minimum-Difficulty-of-a-Job-Schedule/1335.cpp
class Solution {
public:
    int dp(int idx, int d, int maxWork, vector<int>& jd, vector<vector<vector<int>>> &cache) {
        if (idx == jd.size()) {
            if (d > 0) return 1e9;
            return maxWork;
        }
        if (cache[idx][d][maxWork] != -1) {
            return cache[idx][d][maxWork];
        }

        int ans = dp(idx + 1, d, max(maxWork, jd[idx]), jd, cache);
        if (d > 0) {
            ans = min(ans, dp(idx + 1, d - 1, jd[idx], jd, cache) + maxWork);
        }

        return cache[idx][d][maxWork] = ans;
    }
    int minDifficulty(vector<int>& jD, int d) {
        int n = (int)jD.size();
        vector<vector<vector<int>>> cache(n, vector<vector<int>> (d, vector<int>(1001, -1)));
        int ans = dp(1, d - 1, jD[0], jD, cache);
        return ans == 1e9 ? -1 : ans;
    }
};